internal/utils.go

```diff
@@ -72,36 +72,120 @@ func (l *LockMap) Get(k string) string {
-type XSyncMapCounterMap struct {
-	counts *xsync.MapOf[string, *xsync.Counter]
+type state struct {
+	counters *xsync.MapOf[string, *xsync.Counter]
 }
 
-func NewXSyncMapCounterMap() *XSyncMapCounterMap {
-	return &XSyncMapCounterMap{counts: xsync.NewMapOf[string, *xsync.Counter]()}
+func newState() *state {
+	return &state{
+		counters: xsync.NewMapOf[string, *xsync.Counter](),
+	}
+}
+
+func (s *state) reset() {
+	s.counters.Clear()
+}
+
+type CounterMap struct {
+	// Pointer to the active state - allows atomic swapping.
+	activeState atomic.Pointer[state]
+
+	// Prevents concurrent GetAndReset operations.
+	mu *sync.RWMutex
+
+	// Pool for state objects to reduce GC pressure
+	statePool sync.Pool
+
+	inFlightOps *atomic.Int64
+	cond        *sync.Cond
 }
 
-func (cm *XSyncMapCounterMap) Inc(key string) {
-	val, ok := cm.counts.Load(key)
-	if !ok {
-		val, _ = cm.counts.LoadOrStore(key, xsync.NewCounter())
+func NewCounterMap() *CounterMap {
+	cm := &CounterMap{
+		mu: &sync.RWMutex{},
+		statePool: sync.Pool{
+			New: func() any {
+				return newState()
+			},
+		},
+		inFlightOps: &atomic.Int64{},
+		cond:        sync.NewCond(&sync.Mutex{}),
 	}
-	val.Inc()
+	cm.activeState.Store(cm.getStateFromPool())
+	return cm
+}
+
+// getStateFromPool retrieves a state from the pool or creates a new one
+func (cm *CounterMap) getStateFromPool() *state {
+	s := cm.statePool.Get().(*state)
+	s.reset()
+	return s
 }
 
-func (cm *XSyncMapCounterMap) GetAndReset() map[string]int64 {
-	ret := map[string]int64{}
-	cm.counts.Range(func(key string, _ *xsync.Counter) bool {
-		v, ok := cm.counts.LoadAndDelete(key)
-		if ok {
-			ret[key] = v.Value()
+func (cm *CounterMap) Inc(key string) {
+	// Mark operation as in-flight.
+	cm.inFlightOps.Add(1)
+
+	// Get the current active state.
+	state := cm.activeState.Load()
+
+	defer func() {
+		// Remove operation from in-flight list.
+		// If this was the last in-flight operation, signal waiters
+		if cm.inFlightOps.Add(-1) == 0 {
+			cm.cond.L.Lock()
+			cm.cond.Signal()
+			cm.cond.L.Unlock()
+		}
+	}()
+
+	// Try to load existing counter.
+	counter, loaded := state.counters.Load(key)
+	if !loaded {
+		// Create a new counter.
+		counter = xsync.NewCounter()
+		actual, loaded := state.counters.LoadOrStore(key, counter)
+		if loaded {
+			// Another goroutine beat us to it, use their counter.
+			counter = actual
+		}
+	}
+
+	// Increment the counter.
+	counter.Inc()
+}
+
+func (cm *CounterMap) GetAndReset() map[string]int64 {
+	// Ensure only one GetAndReset operation runs at a time.
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+
+	// Atomically swap the states.
+	// Create a new empty state for new increments from the pool.
+	newState := cm.getStateFromPool()
+	oldState := cm.activeState.Swap(newState)
+
+	// Wait for all in-flight operations on the old map to complete.
+	cm.cond.L.Lock()
+	for cm.inFlightOps.Load() > 0 {
+		cm.cond.Wait() // Releases the lock while waiting.
+	}
+	cm.cond.L.Unlock()
```

Comment:
```suggestion
	defer cm.cond.L.Unlock()
```
Now that I think, we shouldn't allow any more writes when we start reporting.